### PR TITLE
Port FractionalMaxPool2d from TH to ATen

### PR DESCRIPTION
**Pull request by chandlerzuo**
Tested:

pytest test/test_nn.py -k Fractional

### CODE REVIEW

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+namespace native {
+namespace {
+
+template <typename scalar_t>
+static std::vector<int> fractional_max_pool2d_generate_intervals(
+  scalar_t sample,
+  int inputSize,
+  int outputSize,
+  int poolSize) {
+  scalar_t alpha = (scalar_t) (inputSize - poolSize) / (scalar_t) (outputSize - 1);
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: nit: prefer C++ `static_cast` over C-style cast

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+  std::vector<int> sequence(outputSize, 0);
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: Nice!

nit: But you immediately fill the elements afterward so there is no need to zero-fill it initially. (If you really want to go to hay on this, you don't even need explicit indexing anymore; you could initialize sequence zero-length, reserve and then emplace elements into it.)

File: aten/src/ATen/native/FractionalMaxPool2d.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/FractionalMaxPool2d.cpp]: (I won't mention this again for other occurrences.)

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+              maxIndex = planeIndex;
+            }
+          }
+        }
+
+        AT_ASSERT(maxVal != -std::numeric_limits<scalar_t>::infinity());
+        AT_ASSERT(maxIndex != -1);
+
+        outputForPlane[h * outputW + w] = maxVal;
+        /* +1 to lua index */
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: this comment is out of date

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+        indicesForPlane[h * outputW + w] = maxIndex;
+      }
+    }
+
+  }
+}
+
+void fractional_max_pool2d_out_cpu_template(
+  at::Tensor const& input_,
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: nit: In this codebase, we use the more conventional `const at::Tensor&` spelling. Best not to force people to have to know how both spellings work :)

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+  int outputW = output_size[0];
+  int outputH = output_size[1];
+  int poolSizeW = pool_size[0];
+  int poolSizeH = pool_size[1];
+
+  /* get contiguous input */
+  auto input = input_.contiguous();
+
+  int ndims = input.ndimension();
+  AT_CHECK((ndims == 3 || ndims == 4),
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: We seem to have lost the is empy check

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+  } else {
+    output.resize_({numBatch, numPlanes, outputH, outputW});
+    /* indices will contain the locations for each output point */
+    indices.resize_({numBatch, numPlanes, outputH, outputW});
+
+    int batch;
+#pragma omp parallel for private(batch)
+    for (batch = 0; batch < numBatch; ++batch) {
+      AT_DISPATCH_FLOATING_TYPES(input.type(), "fractional_max_pool2d", [&] {
+        auto input_data = input.data<scalar_t>();
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: nit: Better to do the pointer retrieval outside of the loop; this pointer retrieval is checked and there's no need to do it inside the batch loop. (This probably doesn't matter in practice; it's just the principle of the matter.)

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+          numPlanes, inputW, inputH, outputW, outputH);
+        }
+      );
+  } else {
+    int batch;
+#pragma omp parallel for private(batch)
+    for (batch = 0; batch < numBatch; ++batch) {
+      AT_DISPATCH_FLOATING_TYPES(
+        input.type(), "fractional_max_pool2d_backward", [&] {
+          auto gradInput_data = gradInput.data<scalar_t>();
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: Pointer retrieval outside of loop applies here too :)

File: aten/src/ATen/native/FractionalMaxPool2d.cpp (hunk 6 above)
Comment (ezyang) [reply aten/src/ATen/native/FractionalMaxPool2d.cpp]: Notes on the matter:

```
Many functions in THNN are parallelized in the following way:

    int64_t batch;
#pragma omp parallel for private(batch)
    for (batch = 0; batch < numBatch; ++batch) {
      THNN_(SpatialFractionalMaxPooling_updateOutput_frame)(
        input->data<scalar_t>() + batch * numPlanes * inputH * inputW,
        output->data<scalar_t>() + batch * numPlanes * outputH * outputW,
        THIndexTensor_(data)(indices) + batch * numPlanes * outputH * outputW,
        randomSamples->data<scalar_t>() + batch * numPlanes * 2,
        numPlanes, inputW, inputH, outputW, outputH, poolSizeW, poolSizeH);
    }
  }

This snippet of code is particularly delicate to port, for a few reasons:

1. You need to do a dispatch before you can retrieve the appropriately typed scalar_t pointers, implying use of the AT_DISPATCH_ macro. However...
2. #pragma omp declarations don't work inside lambdas...
3. You do NOT want to do a dynamic dispatch every time around the batch loop; you should do it once to get into the tight loop.

The recommendation is to make a new helper function to contain the OpenMP pragma, and call it from inside the AT_DISPATCH_ macro. Do the extraction of the pointers once, outside of the loop (since they're same every time around.) An example of this being done is in: https://github.com/pytorch/pytorch/pull/15538/files#diff-3e39c997ceefe5db9d4b51504dc8c91cR92 (h/t @Lin Huang)
```

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+  accscalar_t alpha = (accscalar_t)(inputSize - poolSize) / (accscalar_t) (outputSize - 1);
+  if (index == outputSize - 1) {
+    return inputSize - poolSize;
+  } else {
+    return (int) ((index + sample) * alpha) - (int) (sample * alpha);
+  }
+}
+
+template <typename scalar_t, typename index_t>
+__device__ inline scalar_t* get_ref_by_coord(
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: This file looks quite a bit different than `aten/src/THCUNN/generic/SpatialFractionalMaxPooling.cu` and I am wondering how you went about doing the port. Some notable differences I see:

1. The tensor structs no longer statically know their dimensionality, and use TensorInfo now. There must be some back story to why you chose to use TensorInfo here (I didn't follow the intermediate commits, so I can only guess), but unfortunately it's the wrong choice here. TensorInfo is written in such a way that it can work with arbitrary dimension tensors, and that comes at the cost of efficiency and ease-of-use. We should use the right tool for the right job, and in this case, because this is specifically a 2D kernel, we know the dimensionality statically, and thus, PackedTensorAccessor is the right tool for the job.

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+}
+
+template <typename scalar_t>
+__global__ void fractional_max_pool2d_out_cuda_frame(
+  TensorInfo<scalar_t, int> output,
+  TensorInfo<int64_t, int> indices,
+  TensorInfo<scalar_t, int> input,
+  TensorInfo<scalar_t, int> samples,
+  int poolSizeH, int poolSizeW,
+  int PoolSizeWStatic) {
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: Hmm. Once this is detemplatized, does the loop unrolling still happen? I'm not sure if this function gets inlined to its call site.

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+
+  using accscalar_t = at::acc_type<scalar_t, true>;
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: Help combat boolean blindness, and add a comment saying what the `true` is for :)

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+
+  int ourOutputPoint = threadIdx.x + blockIdx.x * blockDim.x;
+  int plane = blockIdx.y;
+  int batch = blockIdx.z;
+
+  int ndims = output.dims;
+
+  // Each thread generates a specific output point
+  if (ourOutputPoint < output.sizes[ndims - 2] * output.sizes[ndims - 1]) {
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: Err... are you trying to make this code dimension generic? It is laudable goal, but not one I'd attempt on the initial port. (Also, this PR doesn't seem to reap the benefits of doing so, since the enclosing function is still specialized for 2d.)

**Comment by ezyang** [comment]
I stopped reviewing after getting halfway through the changes in `SpatialFractionalMaxPooling_updateOutput`, there are too many modifications to the CUDA code for me to feel comfortable reviewing it without a side-by-side diff. We should work out the TensorInfo/PackedTensorAccessor situation and figure out where to go from there. I think that this code should be re-ported to use PackedTensorAccessor, I'm interested to know what makes this not immediately feasible.

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 11 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: The output tensor can be either 3D or 4D since the batch dimension is optional. In TH implementation, output tensor is upcast to 4D before being passed into this function. Here the output tensor keeps its raw dimension, so ndims is either 3 or 4. (This logic also applies to other places when you see things like "ndims - 1").

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 8 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: Back story: there is an optional batch dimension for both input and output tensor, so input&output can be 3D or 4D. The DeviceTensor has an upcast method that adds the batch dimension if it is missing, coercing everything to be 4D. I did not find the upcast method for PackedTensorAccessor. Instead I found some other PR that used TensorInfo in such cases.

Perhaps I should use "resize_" as a replacement of "upcast"?

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 10 above)
Comment (gchanan) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: even better, we should make the parameter a DeviceType (can be another PR).

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 9 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: I don't know the answer for this one. I believe the behavior is the same as the TH implementation. Do you have some suggestion to test / fix?

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 10 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: Added this to the porting guide.

File: aten/src/ATen/native/FractionalMaxPool2d.cpp (hunk 6 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/FractionalMaxPool2d.cpp]: Thanks for writing this up!

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 8 above)
Comment (ezyang) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: Yeah, so for upcasting, there's a pretty simple trick: do an unsqueeze(0) on the tensor before grabbing the accessor with the correct dimensionality.

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 9 above)
Comment (ezyang) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: The change has performance implications, so there are two ways to check: (1) look at the assembly before and after and see if the loop is unrolled or not; (2) run benchmarks to see if it affected anything or not.

Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: I made an update to template on PoolSizeWStatic. This keeps fidelity to the TH implementation. There is a comment in TH implementation that claims this ensures unroll when de-templated. Benchmark test result seems also confirm this claim: 5.59916s without unroll v.s. 5.34821s with unroll.

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 12]
```diff
@@ -108,45 +90,37 @@ __global__ void fractional_max_pool2d_out_cuda_frame(
     assert(maxVal != at::numeric_limits<scalar_t>::lowest());
     assert(maxIndex != -1);
 
-    auto indices_ref = get_ref_by_coord(indices, ndims,
-      batch, plane, outputH, outputW);
-    auto output_ref = get_ref_by_coord(output, ndims,
-      batch, plane, outputH, outputW);
-    *indices_ref = maxIndex;
-    *output_ref = maxVal;
+    indices[batch][plane][outputH][outputW] = maxIndex;
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: ahhh, much better ;)

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+  IntList pool_size,
+  IntList output_size,
+  const Tensor& randomSamples) {
+  int planeDim = 0;
+  int dimh = 1;
+  int dimw = 2;
+  int numBatch = 1;
+
+  int ndims = input.ndimension();
+  for (int i = 0; i < ndims; i++) {
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: nit: You can just test that numel > 0 to test if something is non-empty

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+
+  /* sizes */
+  int numPlanes = input.size(planeDim);
+  int inputH = input.size(dimh);
+  int inputW = input.size(dimw);
+
+  int outputH = output_size[0];
+  int outputW = output_size[1];
+  int poolSizeH = pool_size[0];
+  int poolSizeW = pool_size[1];
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: This doesn't match the original code...

```
void THNN_(SpatialFractionalMaxPooling_updateOutput)(
           THCState *state,
           THCTensor *input,
           THCTensor *output,
           int outputW, int outputH,
           int poolSizeW, int poolSizeH,
           THCIndexTensor *indices,
           THCTensor *randomSamples)
```

...but I think I agree that the new code is correct and the old code is wrong

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 15]
```diff
@@ -0,0 +1,375 @@
...
+    /* indices will contain the locations for each output point */
+    indices.resize_({numPlanes, outputH, outputW});
+  } else {
+    output.resize_({numBatch, numPlanes, outputH, outputW});
+    indices.resize_({numBatch, numPlanes, outputH, outputW});
+  }
+
+  auto output_ = output;
+  auto input_ = input;
+  auto indices_ = indices;
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: I don't see why these assignments were necessary

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 16]
```diff
@@ -0,0 +1,375 @@
...
+
+  if(ndims == 3) {
+    output_.reshape({1, numPlanes, outputH, outputW});
+    indices_.reshape({1, numPlanes, outputH, outputW});
+    input_.reshape({1, input.size(0), input.size(1), input.size(2)});
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: I suspect you meant to do `output_ = output_.reshape(...)` here. As it stands these lines don't do anything. (Also, use `unsqueeze(0)`, it's less error prone)

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 15 above)
Comment (ezyang) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: (see immediately below)

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 17]
```diff
@@ -0,0 +1,375 @@
...
+  }
+
+  // block is limited to 4 warps
+  // grid handles overflow per each plane
+  int outputPlaneSize = output_.size(2) *
+    output_.size(3);
+  dim3 grid((outputPlaneSize + 127) / 128,
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: nit: Any particular reason you inlined THCCeilDiv here? The inlining makes the intent of the line here less clear

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 18]
```diff
@@ -0,0 +1,375 @@
...
+
+    switch (poolSizeW) {
+      FMP_CASE(2);
+      FMP_CASE(3);
+      FMP_CASE(4);
+      FMP_CASE(5);
+      FMP_CASE(6);
+      FMP_CASE(7);
+      default: FMP(-1);
+    }
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: you lost the check for `cudaGetLastError` here

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 19]
```diff
@@ -0,0 +1,393 @@
...
+
+  int numBatch = 1;
+  int planeDim = 0;
+  int heightDim = 1;
+  int widthDim = 2;
+
+  int outputW = output_size[0];
+  int outputH = output_size[1];
+  int poolSizeW = pool_size[0];
+  int poolSizeH = pool_size[1];
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: Unfortunately I gave bad guidance here. I'm sorry but I gave you bad guidance here. `output_size[0]` is height, `output_size[1]` is width. The THNN ordering here is not the right thing to respect; instead we need to respect the documentation. I wrote some more guidance in the porting guide, which I've reproduced below:

## Height-width argument order

Some kernels in THNN have very strange argument order; for example:

```
void THNN_(SpatialFractionalMaxPooling_updateOutput)(
           THCState *state,
           THCTensor *input,
           THCTensor *output,
           int outputW, int outputH,
           int poolSizeW, int poolSizeH,
           THCIndexTensor *indices,
           THCTensor *randomSamples)
```

outputW comes before outputH, even though conventional NCHW ordering says that height comes before width. *There is code generation logic specifically for THNN that reorders output_sizes to match the THNN kernel; you should manually reorder when you port.* So for example, the above arguments become:

```
void fractional_max_pool2d_out(
  Tensor& output,
  Tensor& indices,
  const Tensor& input,
  IntList pool_size,
  IntList output_size,
  const Tensor& randomSamples) {
  ...
  auto outputH = output[0]; // Notice they are reordered
  auto outputW = output[1];
  auto poolSizeH = pool_size[0];
  auto poolSizeW = pool_size[1];
}
```

The tests are supposed to catch if you've gotten it wrong, but we may have missed some cases, so be vigilant. In general, height comes before width. Here is a full list of dimension offsets (from aten/src/ATen/nn_parse.py), where negative numbers mean that you index from the end of the list:

```
DIMENSION_OFFSET = {
    'width': -1,
    'height': -2,
    'B': 0,
    'C': 1,
    'W': -1,
    'H': -2,
    'T': -3,
    'left': 0,
    'right': 1,
    'top': 2,
    'bottom': 3,
    'front': 4,
    'back': 5,
}
```

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 20]
```diff
@@ -264,26 +260,38 @@ void fractional_max_pool2d_backward_out_cuda_template(
...
+  gradInput.zero_();
+
+  auto gradInput_ = gradInput;
+  auto gradOutput_ = gradOutput;
+  auto indices_ = indices;
+
+  if(ndims == 3) {
+    gradInput_.reshape({1, input.size(0), inputH, inputW});
+    gradOutput_.reshape({1, gradOutput.size(0), outputH, outputW});
+    indices_.reshape({1, indices_.size(0), outputH, outputW});
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: Same bug as above

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu [hunk 21]
```diff
@@ -0,0 +1,375 @@
...
+  // grid handles overflow per each plane
+  int outputPlaneSize = gradOutput_.size(2) *
+    gradOutput_.size(3);
+  dim3 grid((outputPlaneSize + 127) / 128,
+            gradInput_.size(1),
+            gradInput_.size(0));
+  dim3 block(outputPlaneSize > 128 ? 128 : outputPlaneSize);
+
+  auto devIndices = indices.packed_accessor<int64_t, 4>();
+AT_DISPATCH_FLOATING_TYPES_AND_HALF(gradOutput.type(),
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: nit: indentation here is a bit wonky

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 14 above)
Comment (ezyang) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: The old code is correct too, weirdly enough; see https://github.com/pytorch/pytorch/pull/15531#discussion_r245895241

**Comment by ezyang** [comment]
Phew, finished reviewing everything! I think the comments above are the only remaining issues.

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 13 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: done

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 14 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: For this case I followed the doc in torch/nn/modules/pooling.py

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 17 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: I wanted to remove all header files from THNN/THCUNN, so I moved the inline implementation here. I also thought it was quite common in CUDA (it appeared in the elementary CUDA tutorial: https://devblogs.nvidia.com/even-easier-introduction-cuda/).

I am going to add a comment line for readability; let me know if you suggest otherwise.

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 18 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: done

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 20 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: Done.

File: aten/src/ATen/native/cuda/FractionalMaxPool2d.cu (hunk 21 above)
Comment (chandlerzuo) [reply aten/src/ATen/native/cuda/FractionalMaxPool2d.cu]: Fixed.

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 22]
```diff
@@ -0,0 +1,393 @@
...
+  const at::Tensor& randomSamples) {
+
+  int numBatch = 1;
+  int planeDim = 0;
+  int heightDim = 1;
+  int widthDim = 2;
+  int outputW = output_size[0];
+  int outputH = output_size[1];
+  int poolSizeW = pool_size[0];
+  int poolSizeH = pool_size[1];
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: This still doesn't look fixed

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 23]
```diff
@@ -0,0 +1,393 @@
...
+
+Tensor& fractional_max_pool2d_backward_out_cpu(
+  at::Tensor& gradInput,
+  const at::Tensor& gradOutput_,
+  const at::Tensor& input,
+  IntList pool_size,
+  IntList output_size,
+  const at::Tensor& indices)
+{
+  gradInput.resize_as_(input);
```
Comment (ezyang) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: You don't need this one right?

**Review by ezyang — APPROVED** [review]
I am preemptively approving but it looks to me as if there is still a substantive correctness bug in the final diff.

File: aten/src/ATen/native/FractionalMaxPool2d.cpp [hunk 24]
```diff
@@ -0,0 +1,393 @@
+#include "ATen/ATen.h"
```
Comment (gchanan) [inline aten/src/ATen/native/FractionalMaxPool2d.cpp]: we are trying to standardize on <> includes for ATen stuff.

File: test/common_nn.py [hunk 25]
```diff
@@ -3064,7 +3069,7 @@ def test_cuda(self, test_case):
             test_case.assertEqual(cpu_output, gpu_output, self.precision)
 
             # Run backwards on CPU and GPU and compare results
-            for i in range(5):
+            for _ in range(5):
```
Comment (gchanan) [inline test/common_nn.py]: nit: try to avoid changing unrelated lines, do that in a different diff.
